gapic/src/main/com/google/gapid/perfetto/models/Selection.java

```diff
@@ -17,95 +17,140 @@
-      if (selections.length == 1) {
-        return selections[0].buildUi(parent, state);
+      if (selections.size() == 1) {
+        return firstSelection().buildUi(parent, state);
       } else {
         return new MultiSelectionView(parent, selections, state);
       }
     }
 
-    @Override
-    public void mark(State state) {
+    public <Key> Selection<Key> getSelection(Kind<Key> type) {
+      return selections.containsKey(type) ? (Selection<Key>) selections.get(type) :type.EmptySelection();
+    }
+
+    public void markTime(State state) {
       // TODO: is this good enough?
-      if (selections.length == 1) {
-        selections[0].mark(state);
+      if (selections.size() == 1) {
+        firstSelection().markTime(state);
       }
     }
 
-    @Override
     public void zoom(State state) {
       // TODO: handle zooming into multiple selections.
-      if (selections.length == 1) {
-        selections[0].zoom(state);
+      if (selections.size() == 1) {
+        firstSelection().zoom(state);
       }
     }
+
+    private Selection<?> firstSelection() {
+      return selections.firstEntry().getValue();
+    }
   }
 
   public static class CombiningBuilder {
-    private final Map<Comparable<?>, ListenableFuture<Combinable<?>>> selections =
+    private final Map<Kind<?>, ListenableFuture<Combinable<?>>> selections =
         Maps.newTreeMap();
 
     @SuppressWarnings("unchecked")
     public <T extends Combinable<T>> void add(
-        Comparable<?> type, ListenableFuture<Combinable<?>> selection) {
+        Kind<?> type, ListenableFuture<Combinable<?>> selection) {
       selections.merge(type, selection, (f1, f2) -> transformAsync(f1, r1 ->
         transform(f2, r2 -> (((T)r1).combine((T)r2)))));
     }
 
-    public ListenableFuture<? extends Selection> build() {
-      if (selections.size() == 1) {
-        return transform(Iterables.getOnlyElement(selections.values()), Combinable::build);
-      }
-
+    public ListenableFuture<MultiSelection> build() {
       return transform(Futures.allAsList(selections.values()), sels -> {
-        Selection[] res = new Selection[sels.size()];
-        for (int i = 0; i < res.length; i++) {
-          res[i] = sels.get(i).build();
+        Iterator<Kind<?>> keys = selections.keySet().iterator();
+        Iterator<Combinable<?>> vals = sels.iterator();
+        TreeMap<Kind<?>, Selection<?>> res = Maps.newTreeMap();
+        while (keys.hasNext()) {
+          res.put(keys.next(), vals.next().build());
         }
         return new MultiSelection(res);
       });
     }
 
     public static interface Combinable<T extends Combinable<T>> {
       public T combine(T other);
-      public Selection build();
+      public Selection<?> build();
     }
   }
+
+  public static class Kind<Key> implements Comparable<Kind<?>>{
+    public int priority;
+
+    public Kind(int priority) {
+      this.priority = priority;
+    }
+
+    public Selection<Key> EmptySelection() {
```

Comment:
Functions should start with a lowercase function. Mirror `Collections.emptyList()`, which contains just a cast around a static final.